tensorrt_llm/inputs/multimodal.py

```
@@ -0,0 +1,263 @@
+        hasher = hash_lib()
+        if isinstance(image, torch.Tensor):
+            # TODO: Device tensor hashing is an open issue. Limited hashing to CPU for now.
+            image = image.cpu()
+        hasher.update(serialize_item(image))
+        return hasher.hexdigest()
+
+    mm_items = {
+        modality: items if isinstance(items, list) else [items]
+        for modality, items in mm_data.items()
+    }
+    # TODO: need to hash both modality and item to distinguish modality (vllm/PR)
+    mm_hashes = {
+        modality: [_hash_image(item) for item in items]
+        for modality, items in mm_items.items()
+    }
+    return mm_hashes
+
+
+def hexdigest_to_int32(hex_digest: str) -> list[int]:
+    """Convert a 256-bit hexadecimal digest to 8 int32 values."""
+    if len(hex_digest) != 64:
+        raise ValueError(
+            f"Expected 64 character hexadecimal string, got {len(hex_digest)}")
+
+    result = []
+    for i in range(0, 64, 8):
+        hex_chunk = hex_digest[i:i + 8]
+        value = int(hex_chunk, 16)
+        if value > 0x7FFFFFFF:  # Check if the highest bit is set (value > 2^31-1)
+            value = value - 0x100000000  # Convert to signed by subtracting 2^32
+        result.append(value)
+    return result
+
+
+def find_mm_token_lengths(mm_data, input_processor):
+    """Get multimodal token lengths from multimodal data items. """
+
+    mm_items = {
+        modality: items if isinstance(items, list) else [items]
+        for modality, items in mm_data.items()
+    }
+    num_mm_tokens = {}
+
+    for modality, items in mm_items.items():
+        if modality != "image":
+            #TODO: support other modalities
+            return []
+        if not hasattr(input_processor, "get_num_tokens_per_image"):
+            #TODO: backward compatibility for models that don't yet have get_num_tokens_per_image implemented
+            #TODO: only support qwen2_vl for now
```

Should you raise an exception here?